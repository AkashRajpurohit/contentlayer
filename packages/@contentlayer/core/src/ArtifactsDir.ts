import type { OT } from '@contentlayer/utils/effect'
import { pipe, T } from '@contentlayer/utils/effect'
import type { GetContentlayerVersionError } from '@contentlayer/utils/node'
import { fs, getContentlayerVersion } from '@contentlayer/utils/node'
// import utilsPkg from '@contentlayer/utils/package.json'
import * as path from 'path'

export namespace ArtifactsDir {
  export const getDirPath = ({ cwd }: { cwd: string }): string => path.join(cwd, 'node_modules', '.contentlayer')

  export const mkdir = ({ cwd }: { cwd: string }): T.Effect<OT.HasTracer, fs.MkdirError, string> =>
    pipe(T.succeed(getDirPath({ cwd })), T.tap(fs.mkdirp))

  export const getCacheDirPath = ({
    cwd,
  }: {
    cwd: string
  }): T.Effect<OT.HasTracer, GetContentlayerVersionError, string> =>
    pipe(
      getContentlayerVersion(),
      T.map((contentlayerVersion) => path.join(getDirPath({ cwd }), '.cache', `v-${contentlayerVersion}`)),
    )

  export const mkdirCache = ({
    cwd,
  }: {
    cwd: string
  }): T.Effect<OT.HasTracer, fs.MkdirError | GetContentlayerVersionError, string> =>
    pipe(getCacheDirPath({ cwd }), T.tap(fs.mkdirp))
}
