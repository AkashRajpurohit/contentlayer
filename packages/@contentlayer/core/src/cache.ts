import type { E } from '@contentlayer/utils/effect'
import { OT, pipe, T } from '@contentlayer/utils/effect'
import * as node from '@contentlayer/utils/node'
import * as path from 'path'

import type { Document } from './data'
import { getArtifactsDir } from './utils'

export type Cache = {
  /**
   * A map containing all documents wrapped in a {@link CacheItem} indexed by id.
   * We're using a map/record here (instead of a simple array) since it's easier and more efficient
   * to implement cache operations (e.g. lookup, update, delete) this way.
   */
  cacheItemsMap: { [id: string]: CacheItem }
}

export type CacheItem = {
  document: Document
  /**
   * The `documentHash` is used to determine if a document has changed and it's value-generation is implemented
   * by a given plugin (e.g. based on the last-edit date in source-files)
   */
  documentHash: string
}

export const loadPreviousCacheFromDisk = ({
  schemaHash,
}: {
  schemaHash: string
}): T.Effect<OT.HasTracer, node.ReadFileError | node.JsonParseError, Cache | undefined> => {
  const filePath = path.join(getArtifactsDir(), 'cache', `${schemaHash}.json`)
  return pipe(
    node.readFileJson<Cache>(filePath),
    T.catchTag('FileNotFoundError', () => T.succeed(undefined)),
    OT.withSpan('@contentlayer/core/cache:loadPreviousCacheFromDisk', { attributes: { schemaHash } }),
  )
}

export const writeCacheToDisk = ({
  cache,
  schemaHash,
}: {
  cache: Cache
  schemaHash: string
}): T.Effect<OT.HasTracer, never, E.Either<node.WriteFileError | node.MkdirError | node.JsonStringifyError, void>> => {
  const cacheDirPath = path.join(getArtifactsDir(), 'cache')
  const filePath = path.join(cacheDirPath, `${schemaHash}.json`)

  return pipe(
    node.mkdirp(cacheDirPath),
    T.chain(() => node.writeFileJson({ filePath, content: cache })),
    T.either,
  )
}
