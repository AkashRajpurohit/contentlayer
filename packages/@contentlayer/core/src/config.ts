import { SourcePlugin } from './plugin'

export type Config = {
  source: SourcePlugin
}

export function defineConfig(_: Config): Config {
  return _
}
