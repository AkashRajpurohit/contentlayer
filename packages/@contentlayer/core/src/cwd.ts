import type { PosixFilePath } from '@contentlayer/utils'
import { unknownToPosixFilePath } from '@contentlayer/utils'
import type { Has } from '@contentlayer/utils/effect'
import { T, tag } from '@contentlayer/utils/effect'

const CwdSymbol = Symbol()

export const makeCwd = T.gen(function* (_) {
  const cwd = (yield* _(T.succeedWith(() => unknownToPosixFilePath(process.cwd())))) as PosixFilePath

  return { serviceId: CwdSymbol, cwd } as const
})

export interface Cwd extends T.OutputOf<typeof makeCwd> {}
export const Cwd = tag<Cwd>()

export const provideCwd = T.provideServiceM(Cwd)(makeCwd)

export const getCwd = T.accessService(Cwd)((_) => _.cwd)

export type HasCwd = Has<Cwd>
