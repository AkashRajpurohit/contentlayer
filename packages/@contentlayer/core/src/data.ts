export type Document = Record<string, any> & DocumentMeta

export type Object = Record<string, any> & Omit<DocumentMeta, '_id'>

export type DocumentMeta = {
  /**
   * Either coming from API-based CMS or based on the local file path
   * Optional concept as no system/workflow depends on IDs.
   */
  _id: string
  _typeName: string
  _raw?: Record<string, any>
}
