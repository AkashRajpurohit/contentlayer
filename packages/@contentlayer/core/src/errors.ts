import { Tagged } from '@contentlayer/utils/effect'

export class NoConfigFoundError extends Tagged('NoConfigFoundError')<{
  readonly configPath?: string
  readonly cwd: string
}> {
  toString = () =>
    this.configPath
      ? `Couldn't find ${this.configPath}`
      : `Could not find contentlayer.config.ts or contentlayer.config.js in ${this.cwd}`
}

export class ConfigReadError extends Tagged('ConfigReadError')<{
  readonly configPath: string
  readonly error: unknown
}> {}

export class ConfigNoDefaultExportError extends Tagged('ConfigNoDefaultExportError')<{
  readonly configPath: string
  readonly availableExports: string[]
}> {}

export class SourceFetchDataError extends Tagged('SourceFetchDataError')<{
  readonly error: any
}> {
  toString = () => `SourceFetchDataError: ${this.error.toString()}`
}

export class SourceProvideSchemaError extends Tagged('SourceProvideSchemaError')<{
  readonly error: any
}> {
  toString = () => `SourceProvideSchemaError: ${this.error.toString()}`
}
