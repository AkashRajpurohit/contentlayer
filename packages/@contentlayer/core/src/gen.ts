import type { Cache } from './cache'
import type { Document, Object } from './data'

export type GetDocumentTypeMapGen = ContentlayerGen extends { documentTypeMap: infer T } ? T : Record<string, Document>
export type GetDocumentTypeGen<Name extends string> = Name extends keyof GetDocumentTypeMapGen
  ? GetDocumentTypeMapGen[Name]
  : Document
export type GetDocumentTypesGen = ContentlayerGen extends { documentTypes: infer T } ? T : Document
export type GetDocumentTypeNamesGen = ContentlayerGen extends { documentTypeNames: infer T } ? T : string

export type GetObjectTypeMapGen = ContentlayerGen extends { objectTypeMap: infer T } ? T : Record<string, Object>
export type GetObjectTypeGen<Name extends string> = Name extends keyof GetObjectTypeMapGen
  ? GetObjectTypeMapGen[Name]
  : Object
export type GetObjectTypesGen = ContentlayerGen extends { objectTypes: infer T } ? T : Object
export type GetObjectTypeNamesGen = ContentlayerGen extends { objectTypeNames: infer T } ? T : string

export type GetAllTypeNamesGen = ContentlayerGen extends { allTypeNames: infer T } ? T : string

export type GetFieldNamesForDefinitionGen<DefName extends string> = DefName extends keyof GetDocumentTypeMapGen
  ? keyof GetDocumentTypeGen<DefName>
  : keyof GetObjectTypeGen<DefName>

declare global {
  // NOTE will be extended via `node_modules/@types/contentlayer/types/index.d.ts`
  interface ContentlayerGen {}
}

export type CacheGen = Omit<Cache, 'documents'> & { documents: DocumentGen[] }

export type DocumentGen = GetDocumentTypesGen
