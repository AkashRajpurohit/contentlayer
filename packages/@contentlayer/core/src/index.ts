export * from './generation/generate-dotpkg'
export * from './generation/generate-types'
export * from './config'
export * from './DataCache'
export * from './data'
export * from './gen'
export * from './errors'
export * from './getConfig'
export * from './markdown'
export * from './mdx'
export * from './plugin'
export * from './schema'
export * from './utils'
export * from './ArtifactsDir'
