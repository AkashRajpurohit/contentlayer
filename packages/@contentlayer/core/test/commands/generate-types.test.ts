import { defineDocument, fromLocalContent } from 'contentlayer/source-local'
import { buildSource } from '../../src'

export const post = defineDocument(() => ({
  name: 'Post',
  filePathPattern: `**/*.md`,
  fields: {
    title: {
      type: 'string',
      description: 'The title of the post',
      required: true,
    },
    date: {
      type: 'date',
      description: 'The date of the post',
      required: true,
    },
  },
  computedFields: {
    slug: { type: 'string', resolve: (_) => _._id.replace('.md', '') },
  },
}))

describe('generate-types', () => {
  test('simple schema', async () => {
    const schema = await fromLocalContent({ schema: [post], contentDirPath: '' }).provideSchema()
    const typeSource = buildSource(schema)
    expect(typeSource).toMatchSnapshot()
  })
})
