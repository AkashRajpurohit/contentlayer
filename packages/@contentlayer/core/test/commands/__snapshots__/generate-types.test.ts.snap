// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-types simple schema 1`] = `
"// NOTE This file is auto-generated by the Contentlayer CLI
import type { Markdown } from '@contentlayer/core'

export type Image = string
export type { Markdown }



export interface ContentlayerGenTypes {
  documentTypes: DocumentTypes
  documentTypeMap: DocumentTypeMap
  documentTypeNames: DocumentTypeNames
  allTypeNames: AllTypeNames
}

declare global {
  interface ContentlayerGen extends ContentlayerGenTypes {}
}

export type DocumentTypeMap = {
  Post: Post
}

export type AllTypes = DocumentTypes | ObjectTypes
export type AllTypeNames = DocumentTypeNames | ObjectTypeNames

export type DocumentTypes = Post
export type DocumentTypeNames = DocumentTypes['_typeName']

/** Document types */



/** Post */
export type Post = {
  _id: string
  _typeName: 'Post'
  _raw?: Record<string, any>
  /** The title of the post */
  title: string
  /** The date of the post */
  date: string
  /** Default markdown file content */
  content: Markdown | undefined
  slug: string
}

/** Object types */



export type ObjectTypes = never
export type ObjectTypeNames = ObjectTypes['_typeName']


"
`;
